mnist_lenet/__init__.py

```diff
@@ -0,0 +1,250 @@
+
+
+        # Construct a top MLP
+        self.top_mlp = MLP(top_mlp_activations, top_mlp_dims)
+
+        # We need to flatten the output of the last convolutional layer.
+        # This brick accepts a tensor of dimension (batch_size, ...) and
+        # returns a matrix (batch_size, features)
+        self.flattener = Flattener()
+        application_methods = [self.conv_sequence.apply, self.flattener.apply,
+                               self.top_mlp.apply]
+        super(LeNet, self).__init__(application_methods, **kwargs)
+
+    @property
+    def output_dim(self):
+        return self.top_mlp_dims[-1]
+
+    @output_dim.setter
+    def output_dim(self, value):
+        self.top_mlp_dims[-1] = value
+
+    def _push_allocation_config(self):
+        self.conv_sequence._push_allocation_config()
+        conv_out_dim = self.conv_sequence.get_dim('output')
+
+        self.top_mlp.activations = self.top_mlp_activations
+        self.top_mlp.dims = [numpy.prod(conv_out_dim)] + self.top_mlp_dims
+
+
+def main(save_to, num_epochs, feature_maps=None, mlp_hiddens=None,
+         conv_sizes=None, pool_sizes=None, batch_size=500):
+    if feature_maps is None:
+        feature_maps = [20, 50]
+    if mlp_hiddens is None:
+        mlp_hiddens = [500]
+    if conv_sizes is None:
+        conv_sizes = [5, 5]
+    if pool_sizes is None:
+        pool_sizes = [2, 2]
+    image_size = (28, 28)
+    output_size = 10
+
+    # Use ReLUs everywhere and softmax for the final prediction
+    conv_activations = [Rectifier() for _ in feature_maps]
+    mlp_activations = [Rectifier() for _ in mlp_hiddens] + [Softmax()]
+    convnet = LeNet(conv_activations, 1, image_size,
+                    filter_sizes=zip(conv_sizes, conv_sizes),
+                    feature_maps=feature_maps,
+                    pooling_sizes=zip(pool_sizes, pool_sizes),
+                    top_mlp_activations=mlp_activations,
+                    top_mlp_dims=mlp_hiddens + [output_size],
+                    border_mode='full',
+                    weights_init=Uniform(width=.2),
+                    biases_init=Constant(0))
+    # We push initialization config to set different initialization schemes
+    # for convolutional layers.
+    convnet.push_initialization_config()
+    convnet.layers[0].weights_init = Uniform(width=.2)
+    convnet.layers[1].weights_init = Uniform(width=.09)
+    convnet.top_mlp.linear_transformations[0].weights_init = Uniform(width=.08)
+    convnet.top_mlp.linear_transformations[1].weights_init = Uniform(width=.11)
+    convnet.initialize()
+    logging.info("Input dim: {} {} {}".format(
+        *convnet.children[0].get_dim('input_')))
+    for i, layer in enumerate(convnet.layers):
+        logging.info("Layer {} dim: {} {} {}".format(
+            i, *layer.get_dim('output')))
+
+    x = tensor.tensor4('features')
+    y = tensor.lmatrix('targets')
+
+    # Normalize input and apply the convnet
+    probs = convnet.apply((x / 256.) - 0.5)
```

Comment:
If `x` is in [0, 255], I think we should divide by `255.` instead.
Also, is it better to put that kind of normalization as a Transformer in data stream, or in the model directly? Is there a way we would like to encourage more?